### ISSUE TITLE
Inconsistent draining nodes step in Upgrading kubeadm clusters

### ISSUE THREAD

**Issue by sculley** [kubernetes/website#25732]
I noticed that there seems to be inconsistencies in the upgrading kubeadm clusters documentation between 1.19 and 1.20

In the 1.19 [docs](https://v1-19.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrading-control-plane-nodes) after updating kubeadm version and verifying the new version there is a step to drain the pods on the first control-plane node

![Screenshot 2020-12-20 at 12 57 19](https://user-images.githubusercontent.com/3237705/102713868-f5ceee00-42c2-11eb-9930-20b36bcfde16.png)

But in the 1.20 [docs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#call-kubeadm-upgrade) its doesn't list the command to drain the control-plane nodes in the same order

![Screenshot 2020-12-20 at 12 58 32](https://user-images.githubusercontent.com/3237705/102713890-1d25bb00-42c3-11eb-8664-d413e6a4f03f.png)

It does mention draining the control-plane node but not until the section (**For the other control plane nodes**)

![Screenshot 2020-12-20 at 12 59 28](https://user-images.githubusercontent.com/3237705/102713900-37f82f80-42c3-11eb-8bc2-6987254b895b.png)

Also the 1.19 docs don't mention draining any additional control-plane nodes whereas the 1.20 mentions draining the additional nodes

https://v1-19.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-additional-control-plane-nodes

Overall the instructions are not consistent, is this because draining the node is not required before running ```sudo kubeadm upgrade apply v1.20.x``` but is required before upgrading ```kubectl/kubelet``` and restarting ```kubelet```?

Should the docs be updated to ensure the control-plane is drained before applying the upgrade plan as per the instructions in the docs for 1.19?

It's confusing as in the section for upgrading worker nodes, it mentions to drain the node before running the upgrade node command.

![Screenshot 2020-12-20 at 13 40 46](https://user-images.githubusercontent.com/3237705/102714822-43e6f000-42c9-11eb-94e3-96e545e8590f.png)

**Comment by tengqm** [comment]
IIUC, fixes to significant bugs that affect setting up or running Kubernetes are acceptable to archived versions of  the docs.

**Comment by sculley** [comment]
Thanks for the replies, now that we have confirmed draining nodes is required before `kubelet` upgrades and not `kubeadm` I believe that at least 3 PR's will be required;

- Updating [Upgrading Worker Nodes](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-worker-nodes) in the current version of the docs (1.20) to change the position of the [Drain the node](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-worker-nodes) step to be placed after the [Call "kubeadm upgrade"](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-kubelet-and-kubectl-1) step and before the [Upgrade kubelet and kubectl](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-kubelet-and-kubectl-1)

- Subsequent PR's will need to be submitted to modify the archived versions of the docs to ensure the drain/uncordon steps follow the same logic/structure and are applied at the correct stage. (Before upgrading `kubelet` and not `kubeadm`). My only question is how far back do we go I would assume only the last two minor releases (1.19/1.18) as per the Kubernetes project [supported versions](https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-versions)?

If we are in agreement, I can happily work on submitting the PR's I have recently signed the CLA

**Comment by neolit123** [comment]
@tengqm 

> IIUC, fixes to significant bugs that affect setting up or running Kubernetes are acceptable to archived versions of the docs.

i would leave this to the judgment of the docs team. not following the kubelet upgrade recommendations is critical according to SIG Node, yet we haven't seen failure reports by users.

**Comment by neolit123** [comment]
@sculley 

> Updating Upgrading Worker Nodes in the current version of the docs (1.20) to change the position of the Drain the node step to be placed after the Call "kubeadm upgrade" step and before the Upgrade kubelet and kubectl

SGTM, that's one PR for master and one for the dev-1.20 branch.

>  I believe that at least 3 PR's will be required;

one PR per the rest of the dev-x.yy branches (archived releases). once we amend 1.20, such a PR can copy the 1.20 kubeadm-upgrade.md to this dev-x.yy branch.

> My only question is how far back do we go I would assume only the last two minor releases (1.19/1.18) as per the Kubernetes project supported versions?

the documentation goes as far as 1.16, which can be confirmed if you click on the Version dropdown (top right) of k8s.io.
the 1.20 steps are valid for all these versions.

> If we are in agreement, I can happily work on submitting the PR's I have recently signed the CLA

awesome, please keep the PRs enumerated in here.

**Comment by sftim** [comment]
I'm happy to see backports for SIG Cluster Lifecycle topics - especially for critical issues but also otherwise. Essentially, I think it's important that if we ask users to treat the docs for the most recent release as the best version available, then it's important to make sure that any documentation exists to help users / cluster admins get onto that most recent release.

**Comment by sculley** [comment]
> one PR per the rest of the dev-x.yy branches (archived releases). once we amend 1.20, such a PR can copy the 1.20 kubeadm-upgrade.md to this dev-x.yy branch.

@neolit123 - Could you clarify, Why do we need PR's for each of the dev-x.yy branches only? and not the release-x.yy? excuse my ignorance, I've tried searching in Slack channels for an answer to the uses/differences but have had no luck!

And when you say 

>such a PR can copy the 1.20 kubeadm-upgrade.md to this dev-x.yy branch.

What do you mean exactly, the 1.20 kubeadm-upgrade.md file can be used to replace the kubeadm-upgrade.md in all other dev-x.yy branches? Wouldn't that break other things? there would be more changes than just moving the drain the node step around to the correct place?

Many thanks for your assistance.

**Comment by neolit123** [comment]
> Could you clarify, Why do we need PR's for each of the dev-x.yy branches only? and not the release-x.yy? excuse my ignorance, I've tried searching in Slack channels for an answer to the uses/differences but have had no luck!

AFAIK, backports for k/website should be sent to dev-1.xx branches.

> What do you mean exactly, the 1.20 kubeadm-upgrade.md file can be used to replace the kubeadm-upgrade.md in all other dev-x.yy branches?

yes.

>  Wouldn't that break other things? there would be more changes than just moving the drain the node step around to the correct place?

i guess replacing the file completely is almost fine, since the steps for 1.20 are applicable all the way back to 1.16, and more accurate. however, the links on top of the document should be preserved, since they are a bit different.

for example the 1.16 page:
https://v1-16.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/

must still include the old links since those require extra changes:
```
To see information about upgrading clusters created using older versions of kubeadm, please refer to following pages instead:

Upgrading kubeadm cluster from 1.14 to 1.15
Upgrading kubeadm cluster from 1.13 to 1.14
Upgrading kubeadm cluster from 1.12 to 1.13 or Upgrading kubeadm HA clusters from v1.12 to v1.13
```

accordingly for 1.17, 1.18 and 1.19.

**Comment by sftim** [comment]
> AFAIK, backports for k/website should be sent to dev-1.xx branches.

We publish from the _release-1.xx_ branches; as I understand it, the _dev-1.xx_ branches for old releases are vestigial (things we ought to clean up).

I would target the _release-1.xx_ branches for a backport.

**Comment by sculley** [comment]
That's all the PR's submitted, the docs for 1.13 to 1.15 no longer work on the website, so I won't bother with them.

**Comment by tengqm** [comment]
This can be closed when #25887 is in.

**Comment by sculley** [comment]
Thanks!

**Comment by sculley** [comment]
As noticed in #25887 changes made in release-1.16/1.17 made the numbered lists break. The changes were applied in the PR for release-1.16 but the changes for release-1.17 had already been approved and merged in #25886

I need to create a new PR in the release-1.17 branch to fix this issue by updating the numbered lists to be bulleted lists.

**Comment by sculley** [comment]
I believe this can be closed now

**Comment by neolit123** [comment]
thanks for the PRs @sculley  !
/close
